p2p/transport/webrtc/connection.go

```diff
@@ -0,0 +1,344 @@
+	if err != nil {
+		return nil, fmt.Errorf("open stream: %w", err)
+	}
+	stream := newStream(
+		c,
+		dc,
+		rwc,
+		nil,
+		nil,
+	)
+	c.addStream(stream)
+	return stream, nil
+}
+
+func (c *connection) AcceptStream() (network.MuxedStream, error) {
+	select {
+	case <-c.ctx.Done():
+		return nil, os.ErrClosed
+	case accStream := <-c.acceptQueue:
+		stream := newStream(
+			c,
+			accStream.channel,
+			accStream.stream,
+			nil,
+			nil,
+		)
+		c.addStream(stream)
+		return stream, nil
+	}
+}
+
+// implement network.ConnSecurity
+func (c *connection) LocalPeer() peer.ID {
+	return c.localPeer
+}
+
+func (c *connection) LocalPrivateKey() ic.PrivKey {
+	return c.privKey
+}
+
+func (c *connection) RemotePeer() peer.ID {
+	return c.remotePeer
+}
+
+func (c *connection) RemotePublicKey() ic.PubKey {
+	return c.remoteKey
+}
+
+// implement network.ConnMultiaddrs
+func (c *connection) LocalMultiaddr() ma.Multiaddr {
+	return c.localMultiaddr
+}
+
+func (c *connection) RemoteMultiaddr() ma.Multiaddr {
+	return c.remoteMultiaddr
+}
+
+// implement network.ConnScoper
+func (c *connection) Scope() network.ConnScope {
+	return c.scope
+}
+
+func (c *connection) Transport() tpt.Transport {
+	return c.transport
+}
+
+func (c *connection) addStream(stream *webRTCStream) {
+	c.m.Lock()
+	defer c.m.Unlock()
+	c.streams[stream.id] = stream
+}
+
+func (c *connection) removeStream(id uint16) {
+	c.m.Lock()
+	defer c.m.Unlock()
+	delete(c.streams, id)
+}
+
+func (c *connection) onConnectionStateChange(state webrtc.PeerConnectionState) {
+	log.Debugf("[%s][%d] handling peerconnection state: %s", c.localPeer, c.dbgId, state)
+	if state == webrtc.PeerConnectionStateFailed || state == webrtc.PeerConnectionStateClosed {
+		// reset any streams
+		c.resetStreams()
+		c.cancel()
+		// is done safe to call twice?
+		c.scope.Done()
+		c.pc.Close()
+	}
+}
+
+// detachChannel detaches an outgoing channel by taking into account the context
+// passed to `OpenStream` as well the closure of the underlying peerconnection
+func (c *connection) detachChannel(ctx context.Context, dc *webrtc.DataChannel) (rwc datachannel.ReadWriteCloser, err error) {
+	done := make(chan struct{})
+	var cls sync.Once
+	defer cls.Do(func() {
+		close(done)
+	})
+	// OnOpen will return immediately for detached datachannels
+	// refer: https://github.com/pion/webrtc/blob/7ab3174640b3ce15abebc2516a2ca3939b5f105f/datachannel.go#L278-L282
+	dc.OnOpen(func() {
+		rwc, err = dc.Detach()
+		// this is safe since the function should return instantly if the peerconnection is closed
+		cls.Do(func() { close(done) })
+	})
+	select {
+	case <-c.ctx.Done():
+		return nil, errors.New("connection closed")
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case <-done:
+	}
+	return
+}
+
+// only used during connection setup
+func (c *connection) setRemotePeer(id peer.ID) {
+	c.remotePeer = id
+}
+
+// only used during connection setup
+func (c *connection) setRemotePublicKey(key ic.PubKey) {
+	c.remoteKey = key
+}
```

Comment:
Hi @marten-seemann it means I saw your comment and will address it. I was out due to sickness, but I'm getting back on the saddle as we speak, so will address your issue this week!

Reply:
They cannot @marten-seemann as there's an dependency on info from the noise handshake, without having completed the handshake first we cannot set that info. However the handshake might fail if we have not opened the conn, so as such we need to open a partial first and then upgrade.

Within Rust such states (in general) could be solved by providing two types of connections and upgrading one to the other once the remote peer info is available, but for now this is what we have to work with I'm afraid.

Reply:
I'm not convinced. This is called in `transport.dial`, after `transport.noiseHandshake`. This function doesn't even need the `connection`, so resolving this could be as easy as moving the construction of the `connection` after the Noise handshake (which would logically make more sense anyway).

Reply:
@marten-seemann The connection object sets up receiving datachannels (streams) from the remote peer. Please consider the XX noise handshake pattern from a peer A to peer B as described [here](https://noiseexplorer.com/patterns/XX/). 

The initiator A completes the noise handshake before B. This would allow A to create new datachannels before B has set up the callbacks to process incoming datachannels. This would create a situation where A has successfully created a stream but B is not aware of it. Moving the construction of the `connection` object before the noise handshake eliminates this issue, as callbacks have been set up for both peers.

This could lead to a case where streams are created during the noise handshake, and the handshake fails. In this case, we would close the underlying peerconnection.

Reply:
Cool, thanks for the explanation. Can we add something along these lines as a comment? That would be super helpful!

Reply:
@GlenDC ^

Reply:
Added the comment.